sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```
@@ -0,0 +1,427 @@
+ * parse the InputStream into objects usable with Beam.  By default, MongoDbGridFSIO will parse
+ * into Strings, splitting on line breaks and using the uploadDate of the file as the timestamp.
+ */
+public class MongoDbGridFSIO {
+
+  /**
+   * Function for parsing the GridFSDBFile into objects for the PCollection.
+   * @param <T>
+   */
+  public interface ParseCallback<T> extends Serializable {
+    /**
+     * Each value parsed from the file should be output as an
+     * Iterable of Line&lt;T&gt;.  If timestamp is omitted, it will
+     * use the uploadDate of the GridFSDBFile.
+     */
+    public static class Line<T> {
+      final Instant timestamp;
+      final T value;
+
+      public Line(T value, Instant timestamp) {
+        this.value = value;
+        this.timestamp = timestamp;
+      }
+      public Line(T value) {
+        this.value = value;
+        this.timestamp = null;
+      }
+    };
+    public Iterator<Line<T>> parse(GridFSDBFile input) throws IOException;
+  }
+
+  /**
+   * Default implementation for parsing the InputStream to collection of
+   * strings splitting on the cr/lf.
+   */
+  private static class StringsParseCallback implements ParseCallback<String> {
+    static final StringsParseCallback INSTANCE = new StringsParseCallback();
+
+    @Override
+    public Iterator<Line<String>> parse(final GridFSDBFile input) throws IOException {
+      final BufferedReader reader =
+          new BufferedReader(new InputStreamReader(input.getInputStream()));
+      return new Iterator<Line<String>>() {
+        String val = reader.readLine();
+        @Override
+        public boolean hasNext() {
+          return val != null;
+        }
+
+        @Override
+        public Line<String> next() {
+          Line<String> l = new Line<String>(val);
+          try {
+            val = reader.readLine();
+          } catch (IOException e) {
+            val = null;
```

This is incorrect. It will cause the iterator to silently lose all remaining data after any transient IOException. Do not catch the exception - you need to simply rethrow it and let the runner retry the whole bundle.